### ISSUE TITLE
Page View Permission

### ISSUE THREAD

**Issue by wdv4758h** [torchbox/wagtail#257]
At this moment, the published wagtail page can be seen by everyone.

How about a way that can limit the people who can see the specific page (people who want to see that page should login and be in specific role) ?

**Comment by davecranwell** [comment]
Since this issue was first created we've added password protection but it would still be a nice feature to restrict view access to a page, optionally recursively, to a certain user, or to a certain user role/group.

The interface could be similar to Wordpress's approach, where protection could be enabled in one of three ways offered as a set of radio buttons: password (input box for password), specific user (modal to choose user), user group/role (modal to choose group(s))

**Comment by smakinson** [comment]
Has there been any more discussion on this? Its a feature I am looking for. I was thinking about the page privacy modal having a 3rd radio for private to groups and perhaps it would swap out checkboxes with the password field.

**Comment by davecranwell** [comment]
There hasn't, I'm afraid. It's no less on our radar but we prioritise features based on their demand (particularly demand from our clients) so this has the effect of pushing less requested features back.

**Comment by smakinson** [comment]
I am looking into adding this feature. Is vagrant required to run the tests?

**Comment by davecranwell** [comment]
A virtual machine is probably the easiest way for the uninitiated to get up and running but you may have more experience. I don't believe the tests require vagrant at all.

**Comment by kaedroho** [comment]
@smakinson Nope, see my comment on another ticket (https://github.com/torchbox/wagtail/issues/1004#issuecomment-74836009) which shows how to run the tests with just a virtual environment (no postgres/elasticsearch/vagrant) required.

**Comment by smakinson** [comment]
Great, thanks! That testing info looks like what I needed.

**Comment by shacker** [comment]
Looking over the two outstanding PRs re: improving access control for public/private pages, I'm noticing that both of them add a third option based on Users/Group memberships. My org's use case would be a  fourth (and simpler) option, restricting the current hierarchy to logged-in users only (in other words, the site mixes public and private content and needs a way to distinguish access between them). Users may not be in any group at all, so group membership is not quite the right solution.  So, something like:

```
RESTRICTION_CHOICES = (
       (NONE, _("Public")),
       (PASSWORD, _("Private, accessible with the following password")),
       (USERS_GROUPS, _("Private, visible to specific users and groups")),
       (AUTHENTICATED_ONLY, _("Private, visible to any logged-in user"))
   )
```

When a non-auth user accesses a private URL in this case I would expect it to just redirect to the Django LOGIN_URL (which in our case would in turn redirect to a CAS server). No need to display a message to the user in this case, but it would be good to capture the URL fragment so it can be preserved through the redirect.

I could potentially contribute to this PR, but it's not clear which of the two related PRs to base it off of.

**Comment by davecranwell** [comment]
Like that idea. Our agency also does a lot of Drupal, a CMS which comes with "Authenticated Users" as a permission-able usergroup and allows for this kind of setting.

**Comment by gasman** [comment]
I'm marginally in favour of this... on one hand it seems a bit redundant since you can achieve the same thing by creating a group to add the relevant users to, possibly automatically on account creation - and that way you won't have to rework things if you add a 'lesser' category of logged-in user later on (such as 'people who have signed up for email updates'). On the other hand, I'm sure this would be a very common scenario, so on balance it would be good to make the workflow as straightforward as possible.

**Comment by shacker** [comment]
I can work with a workflow that involves adding all users to a default group, and skipping the simple auth/non-auth approach. 

Either way, it would be great to know whether group viewing permissions are on the official wagtail roadmap at this point - I need to know whether to develop my own solution to this or wait for official support. Thanks if you can offer any info there.

**Comment by gasman** [comment]
Hi @shacker - this is indeed on the roadmap as something we'd like to get into Wagtail soon. As far as I'm aware the only thing currently blocking it is dealing with the open questions in https://github.com/torchbox/wagtail/pull/1076#issuecomment-114932870. Basically this boils down to: how should front-end login views work? Are they something that Wagtail provides, or something that developers are expected to implement themselves?

It looks like the two existing PRs take opposing views on this as they stand, and neither option is entirely satisfactory...

Ideally we want group-based view restrictions to work out of the box with no extra setup, which suggests that Wagtail should provide its own basic implementation (as it does for the "password required" prompt for password-based view restrictions). However, it should be possible to override that - both to make it look non-ugly, and allow sites that have their own existing login system to continue using that. But then, if we give developers complete freedom to build their own login views, we can't make assumptions within the Wagtail code about how those views will behave. To give a specific example of such an assumption (taken from my comment linked above): if a user is logged in but has insufficient permission for the page they requested, the nice thing to do is to redirect to the login page with a "you have insufficient permission" message inserted. We could do that through Django's messages framework, but if the developer has complete freedom to customise the login view, can we rely on the relevant blocks being in place on the template? Another example would be redirecting back to the original page after login - can we rely on custom login views to implement that mechanism?

**Comment by shacker** [comment]
I guess the way I would expect it to work would be:

- When editing a page, editors can select a group or groups to which membership is required in order to view (bonus if the implied group of "Authenticated Users" is included in the group list).

- So that developers can choose the behavior when an insufficient permission view is reached, how about a "wt_user_can_view" template tag or similar. Then I can choose either to:

```
{% if wt_user_can_view %}
  content
{% else %}
  sorry you don't permission to view this content
{% endif %}
```

or 

```
{% if wt_user_can_view send_to_login %}
  content
{% endif %}
```

where in the latter case they would be redirected to the site login view. Both approaches would register a Django message, and developers can choose to display Messages or not. 

Just thinking out loud.

**Comment by shacker** [comment]
Maybe that would be cleaner as :

```
{% if wt_user_can_view %}
  content
{% else %}
  {% wt_send_to_login %}
{% endif %}
```

**Comment by SalahAdDin** [comment]
:+1:

**Comment by gasman** [comment]
The template probably isn't the right place for this logic, since we'd usually want to set up the login mechanism once only (or leave it as Wagtail's default) and not have that code repeated on every page type - and I'm not sure if it's even possible to redirect from inside template code (unless you count things like JS redirects, which I'd rather avoid...)

Having said that, we do support per-page-type customisation for the shared-password-based version (http://docs.wagtail.io/en/v1.1/advanced_topics/privacy.html#setting-a-password-required-page-for-a-specific-page-type), so it would be nice to have that as an option here too.

I think what we probably want is to put all of Wagtail's default handlers for these scenarios ("user is not logged in", "user is logged in but has insufficient permissions" - any others?) in a core module (`wagtailcore.views.authentication`, say), along with some bare-bones templates. We can adopt some sensible conventions within those views/templates (e.g. using Django's messages framework to output the "you have insufficient permissions" message) - then depending on how esoteric people's requirements are, they can either leave everything as default (and get ugly unstyled-but-functional login pages), just override the templates, or override the views completely with their own logic. (We'd probably make this configurable in settings - something like `WAGTAIL_AUTH_VIEWS = 'myapp.views.authentication'`.)

**Comment by shacker** [comment]
+1 This sounds like a very smart approach (I was actually thinking later that suggesting that we trigger redirects from within templates was a really bad idea - sorry :)

**Comment by kaedroho** [comment]
I think in most cases, developers would put a link to their main login view in the ``LOGIN_URL`` setting so we could send users there if the developwe hasn't set ``WAGTAIL_AUTH_VIEWS``

**Comment by pyMan** [comment]
I'm interested in user-groups-access too, anyway what about these choices?
- public (as-is today)
- password protected (as-is today)
- only logged user can view (new option)
- only anonymous user can view (new option)

**Comment by gasman** [comment]
Hi @pyMan - That part has already been solved, in both #1076 and #1450. At this point, the only thing blocking getting this into Wagtail is to come up with a general-purpose front-end login system - see the discussion https://github.com/torchbox/wagtail/issues/257#issuecomment-144746216 above.

**Comment by Eraldo** [comment]
I would really like wagtail to be compatible with the "django" way of solving this issue: "Mixins" and "Decorators". That way people can use existing ones like the "LoginRequiredMixin" or write a custom "AccessMixin". That would also solve some future edge-cases where people want something different or specific. And I can for instance decide for myself where to redirect them to.

I am still wondering why there is not yet an "Adapter Class" that enables me to use "Django Classes" with "Pages". (Same goes for ListView, ArchiveView etc) It would be awesome if I could just plug them into my pages.

**Comment by gasman** [comment]
@Eraldo Using mixins/decorators for this would mean that the "private page" behaviour is associated with specific page types, wouldn't it? I don't think that fits with how people expect to use this feature - they should be able to set up password protection on any location on the page tree, regardless of what page type it's using.

**Comment by Eraldo** [comment]
@gasman I was not talking about the 'private page' (password protected) feature. I do see what you mean though. It could be used as well.. 'and' not 'or'. That would also give to option to associate a password protection to a page type.. which could be a wanted use case -- However no the one either of us was talking about as I understand it.

**Comment by gasman** [comment]
@Eraldo Perhaps you could give some code examples of what you have in mind? Right now I can't quite see how it relates to this issue.

**Comment by Eraldo** [comment]
Okay @gasman I read through the conversation tree and admit, that what I suggest is better dealt with in a separate thread.

**Comment by kutenai** [comment]
Is there any update on this one?
Without this feature, it seems as if it is not possible to use wagtail as part of a site that included "logged in" user specific content.

**Comment by gasman** [comment]
Sorry for leaving this hanging for so long. To answer the queries about what's currently blocking this feature:

What I absolutely want to avoid is a setup where - in the default installation - the option is present in the admin but is in a non-usable state. In other words, I want to be sure that the following scenario doesn't happen:

* Alice sets up a new Wagtail site for Bob (or upgrades an existing one), and doesn't pay any special attention to configuring a front-end login mechanism (because that wasn't in the spec)
* Bob browses around the Wagtail admin, discovers the "make this page visible only to users in group X" option, thinks "oh, that's neat" and tries enabling it
* Bob logs out, tries visiting the page, and gets an error page, or a plain "you are not logged in" message with no login form, or is sent to the 'sign in to Wagtail' screen (which is clearly not correct for the typical site visitor who doesn't know or care what Wagtail is)
* Bob complains to Alice about this crappy broken software...

In this scenario, it would be acceptable for Bob to receive a basic unstyled login page. Alternatively, the option could be hidden in the admin until Alice explicitly configures and enables it. (In the latter case, we need to provide documentation about what needs to be configured; in either case, we need to provide documentation for styling the login page.)

From what I can see, none of the existing PRs (#1076, #1450, #2362) currently deal with this in a satisfactory way:

* #1450 and #2362 both provide an "access denied" page with no option to log in at that point
* #1076 provides a basic unstyled login form, but has the issue that it's available at Django's default template path `registration/login.html` - if a developer upgrades an existing site that already has front-end login support (for example, it's a custom Django app that just uses Wagtail for its help pages), and wagtailcore happens to be above that app in INSTALLED_APPS, the existing login page will break on upgrade.

I'm happy to consider any approach that addresses this. Here's my suggestion (feedback and sanity-checking welcomed here!):

* Introduce a `WAGTAIL_FRONTEND_LOGIN_URL` setting, and redirect there when the user needs to authenticate. (Ideally we'd have a separate code path for users who _are_ authenticated but don't have the required permission - but I'm willing to leave that out in the interest of keeping this an MVP...) Anyone who already has a working login mechanism on their site can simply set `WAGTAIL_FRONTEND_LOGIN_URL = LOGIN_URL`.
* Inside [wagtailcore.urls](https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailcore/urls.py), set up a login page at the URL `_util/login`; by default `WAGTAIL_FRONTEND_LOGIN_URL` points here. This can use the standard `django.contrib.auth.views.login` view, with a barebones template at `wagtailcore/login.html`. (Note that we are intentionally _not_ using the Django default URL `/accounts/login/` or the template path `registration/login.html`, to avoid overriding any login mechanism that the project already has in place.)
* To style the login form, a developer can override the template `wagtailcore/login.html` within one of their own apps. For more complex customisations (such as OAuth / social login), they can set up their own login view and point `WAGTAIL_FRONTEND_LOGIN_URL` there. (We could also allow overriding the template on a per-page-type basis, like we do for `password_required_template` - but again that's not really required for an MVP.)

**Comment by ddiazpinto** [comment]
First of all, sorry about my english. ;)

@gasman I'm completely agree with your point of view about this feature and the non-usable state, but being strict, your approach, even if less, have the same problem you have pointed before.

In the scenario that you propose, you have on one hand, a default activated options (like being_login, be_part_of_a_group, password), and some actions that you need to do to configure the workflow behind each them.

An authorization scenario have 3 parts:
 - A resource to be protected
 - An user that makes a request tryin to access to the resource
 - And a workflow that resolves the user's request  `accepting` or `rejecting` it. And, as we are talking about a website, one url endpoint for each result.

So, I think it would be more interesting to define all of it in config. Something like:
```python
WAGTAIL_PAGE_AUTHORIZATION  {
    'authentication_method' : [
        'path.to.first.worlkflow.to.check',
        'path.to.second.workflow.if.first.fails',
        '...'
    ],
    'another_authentication_method': [
        '...',
    ]
}
```
In page configuration we can only choose one of the methods defined before.

In summary, we define a list of authorizations, that will be processed one by one until one of them be accepted or, the authorization fails.

And obviously those workflows would have to provide something like this:

```python
class IsLoggedAuthorization(AbstractAuthorization):
    def is_authorized(request):
        # Logic to check if the request/user have authorized

    def authorized_endpoint(request):
        # Endpoint to the resource, based on the request

    def non_authorized_endpoint(request):
        # Here, depending of the request, you can point the user to a step where to request more information: password, login, etc.
```

That would be my approach.

**Comment by gasman** [comment]
Hi @ddiazpinto - thanks for the suggestion!

I can see why a fully configurable, modular permission system is an attractive idea, but it's not what we need right now. The PRs here are fundamentally working - they just need packaging up into a format with better developer experience, and I'm keen to do that in the most direct, immediate way possible. This is a very frequently requested feature, and it's already been delayed by many months.

What you're proposing would be a complete rewrite of the permission system, probably requiring several months of specification and design, and it's not really clear to me that the end result would be a better experience for users and developers. (Keep in mind that the site editors who want to set up permission restrictions are not necessarily developers - they need to be able to set them up through the admin UI, not by editing a config file.)

**Comment by ddiazpinto** [comment]
@gasman You're welcome ;)

I think that modularity and flexibility are always better for developers and even admin users, but I understand that my approach takes much more time and the necessary to let this feature resolved as soon as possible.

Anyway, one solution for the last thing you said at the end, that normally the people who sets permissions are not developers, is as simple as provide a set of pre-defined authorization methods according to a bare installation of django/wagtail (`AllUsersAuthorization`, `IsUserAdminAuthorization`, `IsUserModeratorAuthorization`, etc...) and some other methods with no external dependencies (like `PasswordAuthorization`), and even some other methods to link with well known django apps, like allauth for example to handle all the authentication stuff (disabled by default obviously). So this way, it is possible to limit the need of configuration even at the first time.
Being right, if to create the first page in wagtail you need to make a model (and some other developer stuff), it is not crazy to think that for make the first bundle of permissions, you are going to need to edit the configuration.

Anyway, as I said, I understand pretty well that it is necessary to finish this feature quickly but, it is possible to let this talk open for future, in other issue or something?

As always, thanks for this amazing app ;)

**Comment by m1kola** [comment]
Implemented in #3071

**Comment by begonaalvarezd** [comment]
Hello all! Any clue when #3071 is going to be released?

**Comment by gasman** [comment]
@begonaalvarezd It will be in version 1.8 - a release candidate will be coming in the next few days, with the final release some time in mid-December.

**Comment by Adrian-Turjak** [comment]
@gasman I know this issue is closed, but before I open a new one for something potentially related to this, is there a way to make a page type that **has** to be private? As in, you can't publish it without setting it to private.

I have a page type that by default exposes data from an internal API that we need to expose to a third-party with a password, but I want to make it VERY hard if not impossible for someone to accidentally expose that page type to everyone.

**Comment by loicteixeira** [comment]
Page visibility settings are saved as a [PageViewRestriction](https://github.com/wagtail/wagtail/blob/a2d41ef6bf35554e5130f01327b571ad909380ce/wagtail/core/models.py#L1899-L1908) object.

If you want to automatically set a new restriction, I imagine you could override the `save` method of your page object to create/update a PageViewRestriction object associated with the said page.
